### PR TITLE
feat: Add SentryMessage

### PR DESCRIPTION
**Pull request by philipphofmann**
## :scroll: Description

Add type SentryMessage for the event.message. SentryMessage can carry a format
string and structured parameters. This can help to group similar messages into
the same issue.

Fixes https://github.com/getsentry/sentry-cocoa/issues/747

## :bulb: Motivation and Context

We want to support this [new type](https://develop.sentry.dev/sdk/event-payloads/message/).

## :green_heart: How did you test it?
Tavis and simulator.

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [x] I updated the CHANGELOG
- [x] I updated the docs if needed
- [ ] No breaking changes

### CODE REVIEW

File: Sources/Sentry/SentryClient.m [hunk 1]
```diff
@@ -159,7 +160,8 @@ - (SentryId *)captureError:(NSError *)error
 - (SentryEvent *)buildErrorEvent:(NSError *)error
 {
     SentryEvent *event = [[SentryEvent alloc] initWithLevel:kSentryLevelError];
-    event.message = [NSString stringWithFormat:@"%@ %ld", error.domain, (long)error.code];
+    NSString *message = [NSString stringWithFormat:@"%@ %ld", error.domain, (long)error.code];
+    event.message = [SentryMessage messageWithFormatted:message];
```
Comment (marandaneto) [inline Sources/Sentry/SentryClient.m:164]: L: wondering if here would be a good case to use the params instead of the formatted field.

Comment (philipphofmann) [reply Sources/Sentry/SentryClient.m:164]: That's very good input thanks 👏 . I will take a look.

File: Tests/SentryTests/Protocol/SentryMessageTests.swift [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        message.formatted = "Something %s"
+        message.message = "A message"
+        message.params = ["my", "params"]
+        
+        let actual = message.serialize()
+        
+        XCTAssertEqual(message.formatted, actual["formatted"] as? String)
+        XCTAssertEqual(message.message, actual["message"] as? String)
+        XCTAssertEqual(message.params, actual["params"] as? [String])
+    }
```
Comment (marandaneto) [inline Tests/SentryTests/Protocol/SentryMessageTests.swift:53]: H: not really related to this test but think about the use case:
the device has a cached event with the old message field called `message` and its a `String`, the new version of the App has the upgraded SDK and now `message` is a type, how would the serialization behave? does it crash? does it callback and set the string to the formatted field? does it fail and delete the event altogether?

File: Sources/Sentry/SentryClient.m (hunk 1 above)
Comment (marandaneto) [reply Sources/Sentry/SentryClient.m:164]: the downside is that every message will be grouped independently of the domain and error code, not sure if its a good idea, but attach stack traces are enabled by default so it won't group by messages unless it's disabled, right?

Comment (philipphofmann) [reply Sources/Sentry/SentryClient.m]: It is anyway going to be grouped into the same issue if just the error code is different. The advantage of this approach is that you see the different error codes in the message section of the issue. With the previous version, you just saw the error code of one issue and you didn't see if it changed.

File: Tests/SentryTests/Protocol/SentryMessageTests.swift (hunk 2 above)
Comment (philipphofmann) [reply Tests/SentryTests/Protocol/SentryMessageTests.swift:53]: I like it a lot that you think about such edge cases 👏 . Until now we only serialize an event and don't deserialize it. So we just store it as an envelope to disk and never look into the envelope item containing the event. Furthermore, we just remove events cached to [disk before version 6.x](https://github.com/getsentry/sentry-cocoa/blob/master/Sources/Sentry/SentryFileManager.m#L64-L66). So we are good.

Comment (philipphofmann) [reply Tests/SentryTests/Protocol/SentryMessageTests.swift:53]: I'm going to add the information to the migration guide tough, that we just delete cached events before v 6.x.

**Review by bruno-garcia — APPROVED** [review]

**Review by marandaneto — APPROVED** [review]
thansk for that, LGTM

File: Sources/Sentry/SentryClient.m (hunk 1 above)
Comment (marandaneto) [reply Sources/Sentry/SentryClient.m]: nice, thanks!

File: Tests/SentryTests/Protocol/SentryMessageTests.swift (hunk 2 above)
Comment (marandaneto) [reply Tests/SentryTests/Protocol/SentryMessageTests.swift:53]: nice, thanks

**Pull request merged**